typescript/infra/src/funding/grafana.ts

```diff
@@ -0,0 +1,207 @@
+    case WalletName.Relayer:
+      return /wallet_name="relayer", chain="([^"]+)"[^-]+ - ([0-9.]+)/g;
+    default:
+      throw new Error(`Unknown wallet name: ${walletName}`);
+  }
+}
+
+export async function getAlertThresholds(
+  alertType: AlertType,
+): Promise<ChainMap<string>> {
+  const saToken = await fetchServiceAccountToken();
+  const alert = await fetchGrafanaAlert(alertType, saToken);
+  const alertQuery = alert.queries[0];
+  const walletName = alertConfigMapping[alertType].walletName;
+  return parsePromQLQuery(alertQuery, walletName);
+}
+
+export async function fetchServiceAccountToken(): Promise<string> {
+  let saToken: string | undefined;
+
+  try {
+    saToken = (await fetchGCPSecret(
+      'grafana-balance-alert-thresholds-token',
+      false,
+    )) as string;
+  } catch (error) {
+    logger.error(
+      'Error fetching grafana service account token from GCP secrets:',
+      error,
+    );
+    throw error;
+  }
+
+  return saToken;
+}
+
+export async function updateGrafanaAlert(
+  alertUid: string,
+  existingAlert: ProvisionedAlertRule,
+  newQuery: string,
+  saToken: string,
+) {
+  // Create the updated rule based on the existing one
+  const updatedRule: ProvisionedAlertRule = {
+    ...existingAlert,
+    data: existingAlert.data.map((d) => ({
+      ...d,
+      model: {
+        ...d.model,
+        expr: newQuery,
+      },
+    })),
+  };
+
+  const response = await fetch(
+    `${GRAFANA_URL}/api/v1/provisioning/alert-rules/${alertUid}`,
+    {
+      method: 'PUT',
+      headers: {
+        Authorization: `Bearer ${saToken}`,
+        'Content-Type': 'application/json',
+        'X-Disable-Provenance': 'true',
+      },
+      body: JSON.stringify(updatedRule),
+    },
+  );
+
+  if (!response.ok) {
+    const errorData = await response.json();
+    throw new Error(
+      `Failed to update alert: ${response.status} ${JSON.stringify(errorData)}`,
+    );
+  }
+
+  return response.json();
+}
+
+export function generateQuery(
+  alertType: AlertType,
+  thresholds: ChainMap<string>,
+): string {
+  const config = alertConfigMapping[alertType];
+  const walletQueryName = walletNameQueryFormat[config.walletName];
+
+  // TODO: abstract away special handling for relayer queries that need hyperlane_context
+  const needsHyperlaneContext = config.walletName === WalletName.Relayer;
+
+  const queryFragments = Object.entries(thresholds).map(
+    ([chain, minBalance]) => {
+      const labels = [`wallet_name="${walletQueryName}"`, `chain="${chain}"`];
+      if (needsHyperlaneContext) {
+        labels.push('hyperlane_context="hyperlane"');
+      }
+      return `last_over_time(hyperlane_wallet_balance{${labels.join(
+        ', ',
+      )}}[1d]) - ${minBalance} or`;
+    },
+  );
+
+  return `${config.queryTemplate.header}
+    ${queryFragments.join('\n    ')}
+
+${config.queryTemplate.footer}
+)`;
+}
+
+export function sortThresholds(
```

Comment:
whoa I had no idea Javascript objects had a stable ordering of keys following the insertion order, neat. Maybe this is better in `utils/objects.ts` as a generic `sortObject` or something? with a comment why it works, imo it's not obvious